crates/sui-core/src/execution_cache/proxy_cache.rs

```diff
@@ -0,0 +1,422 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::authority::authority_store::{ExecutionLockWriteGuard, SuiLockResult};
+use crate::authority::epoch_start_configuration::EpochStartConfiguration;
+use crate::authority::epoch_start_configuration::{EpochFlag, EpochStartConfigTrait};
+use crate::authority::AuthorityStore;
+use crate::state_accumulator::AccumulatorStore;
+use crate::transaction_outputs::TransactionOutputs;
+
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use parking_lot::Mutex;
+use std::sync::Arc;
+use std::time::Duration;
+use sui_protocol_config::ProtocolVersion;
+use sui_types::accumulator::Accumulator;
+use sui_types::base_types::VerifiedExecutionData;
+use sui_types::base_types::{EpochId, ObjectID, ObjectRef, SequenceNumber};
+use sui_types::bridge::Bridge;
+use sui_types::digests::{TransactionDigest, TransactionEffectsDigest, TransactionEventsDigest};
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use sui_types::object::Object;
+use sui_types::storage::{MarkerValue, ObjectKey, ObjectOrTombstone, PackageObject};
+use sui_types::sui_system_state::SuiSystemState;
+use sui_types::transaction::{VerifiedSignedTransaction, VerifiedTransaction};
+
+use super::{
+    CheckpointCache, ExecutionCacheCommit, ExecutionCacheConfigType, ExecutionCacheMetrics,
+    ExecutionCacheReconfigAPI, ExecutionCacheWrite, ObjectCacheRead, PassthroughCache,
+    StateSyncAPI, TestingAPI, TransactionCacheRead, WritebackCache,
+};
+
+macro_rules! delegate_method {
+    ($self:ident.$method:ident($($args:ident),*)) => {
+        match *$self.mode.lock() {
+            ExecutionCacheConfigType::PassthroughCache => $self.passthrough_cache.$method($($args),*),
+            ExecutionCacheConfigType::WritebackCache => $self.writeback_cache.$method($($args),*),
+        }
+    };
+}
+
+pub struct ProxyCache {
+    // Note: both caches must be constructed at startup, rather than using ArcSwap
+    // (or some similar strategy). This is because we need to proxy iter_live_object_set,
+    // which requires that we borrow from a member of the cache. If we used ArcSwap,
+    // we would be forced to borrow from a local variable after loading from the ArcSwap.
+    //
+    // Cache implementations are entirely passive, so the unused one will have no effect.
+    passthrough_cache: PassthroughCache,
+    writeback_cache: WritebackCache,
+    mode: Mutex<ExecutionCacheConfigType>,
```

Comment:
Well, sort of. We copy the mode out of the lock before calling the proxied method, so the temporary should be dropped immediately, so the critsec should be extremely short. But a RwLock is almost certainly better.

I ran the baseline benchmark and the RwLock is definitely faster, although what is interesting is that the ProxyCache seems to have a huge amount of overhead, it's basically 2x slower. That's ok, since this will get deleted once we are fully migrated, and it's more than fast enough for now. But still pretty interesting.